### PR TITLE
Provide an E2E test for the OSBA.

### PR DESCRIPTION
**Pull request by y-cote**
Add a complete e2e test for the selection and validation of an app selected from the OSBA.

```release-note
NONE
```

### CODE REVIEW

**Comment by asalkeld** [comment]
looking good

File: test/e2e/specs/end_user.go [hunk 1]
```diff
@@ -30,3 +30,25 @@ var _ = Describe("Openshift on Azure end user e2e tests [EndUser][Fake]", func()
 		Expect(errs).To(BeEmpty())
 	})
 })
+
+var _ = Describe("Openshift on Azure end user e2e tests [EndUser][Fake]", func() {
```
Comment (asalkeld) [inline test/e2e/specs/end_user.go]: shouldn't this have a different description to the above one?

File: Makefile [hunk 2]
```diff
@@ -171,6 +171,9 @@ codecov: unit
 e2e:
 	FOCUS="\[CustomerAdmin\]|\[EndUser\]\[Fake\]" TIMEOUT=60m ./hack/e2e.sh
 
+e2e-osba:
+	FOCUS="\[OSBA\]\[Apps\]" TIMEOUT=60m ./hack/e2e.sh
```
Comment (jim-minter) [inline Makefile]: not needed - mark the test [EndUser]

File: pkg/vmimage/bindata.go [hunk 3]
```diff
@@ -67,7 +67,7 @@ func (fi bindataFileInfo) Sys() interface{} {
...
...\x40\xa7\xb3\xc9\x64\x73\x5b\xf3\x86\xec\xc3\x07\xcc\xc6\xf9\x09\x47\x03\xcf\x61\x0c\x27\x27\x3f\xa1\x55\x6f\x08\x73\x59\xe2\x58\xd4\xbb\xff\x2d\xe9\xa0\xc2\x34\x28\xda\xbd\xf3\x70\xb6\x71\xe3\x4a\xac\x38\x4b\xe5\xbc\x17\xd3\x2e\xdf\x4d\xb0\xe7\x65\x6d\xa4\xc1\x93\xca\x72\x76\xe5\x62\x1f\x7a\x9a\x16\x27\x62\xe8\x5d\x0c\xa2\x94\xdb\xc8\x8c\xec\x8d\x2d\x4c\x48\x8c\xf8\xb6\xe1\xb7\x8b\xc9\x38\x7f\x7c\xfc\xe7\xe3\xe0\xb1\xfb\xf8\xd5\xe3\x37\x8f\x8f\xb7\x35\x14\x21\x86\xf3\xe9\x2d\xb0\x45\x59\x19\x1d\x9c\xcd\xd1\x27\x06\x9d\xb9\xc0\xb9\x1e\xe8\xe9\xe7\xff\x16\x2c\xf3\xb3\x2a\xf4\x59\x2f\xb9\xa8\x26\x9c\x71\x54\xab\xbc\x96\x03\x6d\x5d\xfe\xab\x6f\x3b\xe2\x42\x38\x2b\xc3\xb1\x4c\xfe\x5b\xd6\xdc\xf7\xff\x9d\x4e\xe3\xff\xb9\x16\x98\x93\xff\x22\x8f\x58\xec\xc5\x62\xc8\x3d\x1f\x79\x11\xfa\xe0\x83\x14\x92\xf9\xc1\x42\xb2\xe3\xc7\x9d\xe0\x52\xd2\x2b\x60\x84\xc0\x80\x4f\x2f\x67\xfe\x41\x52\x16\x65\xd8\x94\x77\xe1\x23\x0c\xc3\xf8\x4e\xa0\x45\x19\x9a\x8c\xf1\x64\x44\x09\x40\x52\x86\x99\xbb\x62\x9e\xa6\x82\xc2\x45\xaa\x64\x5a\x28\x03\x4d\xf6\x7a\xc8\xc2\x88\x52\x28\x7b\x0e\x03\x2e\x91\xb2\x31\x3f\x17\xec\x54\x08\x8c\xf3\xe8\x24\x82\x1c\x28\x2e\xa1\xd3\x59\x19\xd8\x51\x50\xc4\x00\xd0\x25\x64\x4f\x71
```
Comment (jim-minter) [inline pkg/vmimage/bindata.go]: revert changes to pkg/vmimage

File: test/manifests/osba/osba-os-template.yaml [hunk 4]
```diff
@@ -0,0 +1,237 @@
+kind: Template
```
Comment (jim-minter) [inline test/manifests/osba/osba-os-template.yaml]: can put a yaml comment saying where this file comes from

File: test/e2e/standard/osbachecker.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+	if os.IsNotExist(err) {
+		osbaService.data, err = ioutil.ReadFile("test/e2e/manifests/osba/osba-os-template.yaml") // running via compiled test binary
+	}
+	if err != nil {
+		panic(err)
+	}
+}
+
+// ValidateOSBAApp verifies that the created application runs normally
+func (sc *SanityChecker) ValidateOSBAApp(ctx context.Context, cookie interface{}) []*TestError {
```
Comment (jim-minter) [inline test/e2e/standard/osbachecker.go]: Do not extend SanityChecker for this.  Put all these functions in or next to test/e2e/specs/osba_app.go.

File: test/clients/openshift/template.go [hunk 6]
```diff
@@ -1,15 +1,29 @@
...
 	"github.com/openshift/openshift-azure/pkg/util/ready"
 )
 
+func init() {
+	// The Kubernetes Go client (nested within the OpenShift Go client)
+	// automatically registers its types in scheme.Scheme, however the
+	// additional OpenShift types must be registered manually.  AddToScheme
+	// registers the API group types (e.g. route.openshift.io/v1, Route) only.
+
+	templatev1.AddToScheme(scheme.Scheme)
```
Comment (jim-minter) [inline test/clients/openshift/template.go]: s/AddToScheme/AddToSchemeInCoreGroup/

File: test/clients/openshift/template.go [hunk 7]
```diff
@@ -31,3 +45,46 @@ func (cli *Client) InstantiateTemplate(srcTemplateName, dstNamespace string) err
 
 	return wait.PollImmediate(2*time.Second, 10*time.Minute, ready.CheckTemplateInstanceIsReady(cli.TemplateV1.TemplateInstances(dstNamespace), dstNamespace))
 }
+
+// InstantiateTemplateFromBytes instantiates an openshift template from a byte slice
+func (cli *Client) InstantiateTemplateFromBytes(yamlData []byte, dstNamespace string) error {
+	s := json.NewYAMLSerializer(json.DefaultMetaFactory, scheme.Scheme, scheme.Scheme)
+
+	// Decode the YAML to an object.
+	var list corev1.List
```
Comment (jim-minter) [inline test/clients/openshift/template.go]: you're decoding a Template, not a list: `var template templatev1.Template`

File: test/clients/openshift/template.go [hunk 8]
```diff
@@ -31,3 +45,46 @@ func (cli *Client) InstantiateTemplate(srcTemplateName, dstNamespace string) err
...
+	_, _, err := s.Decode(yamlData, nil, &list)
+	if err != nil {
+		return err
+	}
+
+	// Some types, e.g. List, contain RawExtensions.  If the appropriate types
+	// are registered, these can be decoded in a second pass.
+	for i, o := range list.Items {
```
Comment (jim-minter) [inline test/clients/openshift/template.go]: none of this is needed - you should be able to post the template you unmarshal verbatim

File: test/e2e/standard/osbachecker.go [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+		errs = append(errs, &TestError{Err: err, Bucket: "DeleteOSBAApp"})
+	}
+	return errs
+}
+
+// CreateOSBAApp creates the OSBA Application under test
+func (sc *SanityChecker) CreateOSBAApp(ctx context.Context) (interface{}, []*TestError) {
+	var errs []*TestError
+	sc.log.Debugf("creating openshift project for broker app")
+	template, err := random.LowerCaseAlphanumericString(5)
```
Comment (jim-minter) [inline test/e2e/standard/osbachecker.go]: for better or worse, currently the template is hard-coded such that it only works if the namespace is `osba`.  This code can go.

File: test/clients/openshift/template.go [hunk 10]
```diff
@@ -4,12 +4,29 @@ import (
...
 	"github.com/openshift/openshift-azure/pkg/util/ready"
 )
 
+func init() {
+	// The Kubernetes Go client (nested within the OpenShift Go client)
+	// automatically registers its types in scheme.Scheme, however the
+	// additional OpenShift types must be registered manually.  AddToScheme
+	// registers the API group types (e.g. route.openshift.io/v1, Route) only.
+
+	templatev1.AddToSchemeInCoreGroup(scheme.Scheme)
```
Comment (jim-minter) [inline test/clients/openshift/template.go]: best to check the error and panic if != nil

File: test/e2e/specs/osba_app.go [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+)
+
+type osbaObj struct {
+	namespace string
+	data      []byte
+}
+
+var osbaApp, osbaService osbaObj
+
+func init() {
```
Comment (jim-minter) [inline test/e2e/specs/osba_app.go]: don't think any of this really should be done in init(), nor are the globals needed

File: test/e2e/specs/osba_app.go [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+
+// ValidateOSBAApp verifies that the created application runs normally
+func validateOSBAApp(ctx context.Context, cookie interface{}) []*standard.TestError {
+	var errs []*standard.TestError
+
+	sanity.Checker.Log.Debugf("validating that all osba components are healthy")
+	err := wait.Poll(2*time.Second, 20*time.Minute, ready.CheckClusterServiceBrokerIsReady(sanity.Checker.Client.CustomerAdmin.ServicecatalogV1beta1.ClusterServiceBrokers(), osbaService.namespace))
+	if err != nil {
+		sanity.Checker.Log.Error(err)
+		errs = append(errs, &standard.TestError{Err: err, Bucket: "ready.CheckClusterServiceBrokerIsReady"})
```
Comment (jim-minter) [inline test/e2e/specs/osba_app.go]: (the point of the buckets is that they are few - "validateOSBAApp" or even just "osba" is better - but prefer to return error)

File: test/e2e/specs/osba_app.go [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+	if os.IsNotExist(err) {
+		osbaService.data, err = ioutil.ReadFile("test/e2e/manifests/osba/osba-os-template.yaml") // running via compiled test binary
+	}
+	if err != nil {
+		panic(err)
+	}
+}
+
+// ValidateOSBAApp verifies that the created application runs normally
+func validateOSBAApp(ctx context.Context, cookie interface{}) []*standard.TestError {
```
Comment (jim-minter) [inline test/e2e/specs/osba_app.go]: any good reason not to return `error` throughout this file instead of `[]*standard.TestError`?

Comment (jim-minter) [reply test/e2e/specs/osba_app.go]: don't try to fit into the sanity checker style - no need for cookie interface{} in any of this file

File: test/e2e/specs/osba_app.go [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+	sanity.Checker.Log.Debugf("creating openshift broker app in %s", osbaService.namespace)
+	sanity.Checker.Log.Debugf("instantiating template for %s", osbaService.namespace)
+	err = sanity.Checker.Client.CustomerAdmin.InstantiateTemplateFromBytes(osbaService.data, osbaService.namespace, parameters)
+	if err != nil {
+		sanity.Checker.Log.Error(err)
+		errs = append(errs, &standard.TestError{Err: err, Bucket: "sanity.Checker.Client.CustomerAdmin.InstantiateTemplateFromBytes"})
+		return nil, errs
+	}
+
+	/*
```
Comment (jim-minter) [inline test/e2e/specs/osba_app.go]: remove

**Comment by jim-minter** [comment]
I think this is looping on a stale test (v4.0) - otherwise I think tests are passing - merging maually

**Pull request merged**